lib/imw/rip/wget.man
WGET(1)                            GNU Wget                            WGET(1)



NAME
       Wget - The non-interactive network downloader.

SYNOPSIS
       wget [option]... [URL]...

DESCRIPTION
       GNU Wget is a free utility for non-interactive download of files from
       the Web.  It supports HTTP, HTTPS, and FTP protocols, as well as
       retrieval through HTTP proxies.

       Wget is non-interactive, meaning that it can work in the background,
       while the user is not logged on.  This allows you to start a retrieval
       and disconnect from the system, letting Wget finish the work.  By con‐
       trast, most of the Web browsers require constant user’s presence, which
       can be a great hindrance when transferring a lot of data.

       Wget can follow links in HTML and XHTML pages and create local versions
       of remote web sites, fully recreating the directory structure of the
       original site.  This is sometimes referred to as ‘‘recursive download‐
       ing.’’  While doing that, Wget respects the Robot Exclusion Standard
       (/robots.txt).  Wget can be instructed to convert the links in down‐
       loaded HTML files to the local files for offline viewing.

       Wget has been designed for robustness over slow or unstable network
       connections; if a download fails due to a network problem, it will keep
       retrying until the whole file has been retrieved.  If the server sup‐
       ports regetting, it will instruct the server to continue the download
       from where it left off.

OPTIONS
       Option Syntax

       Since Wget uses GNU getopt to process command-line arguments, every
       option has a long form along with the short one.  Long options are more
       convenient to remember, but take time to type.  You may freely mix dif‐
       ferent option styles, or specify options after the command-line argu‐
       ments.  Thus you may write:

               wget -r --tries=10 http://fly.srk.fer.hr/ -o log

       The space between the option accepting an argument and the argument may
       be omitted.  Instead -o log you can write -olog.

       You may put several options that do not require arguments together,
       like:

               wget -drc <URL>

       This is a complete equivalent of:

               wget -d -r -c <URL>

       Since the options can be specified after the arguments, you may termi‐
       nate them with --.  So the following will try to download URL -x,
       reporting failure to log:

               wget -o log -- -x

       The options that accept comma-separated lists all respect the conven‐
       tion that specifying an empty list clears its value.  This can be use‐
       ful to clear the .wgetrc settings.  For instance, if your .wgetrc sets
       "exclude_directories" to /cgi-bin, the following example will first
       reset it, and then set it to exclude /~nobody and /~somebody.  You can
       also clear the lists in .wgetrc.

               wget -X ’’ -X /~nobody,/~somebody

       Most options that do not accept arguments are boolean options, so named
       because their state can be captured with a yes-or-no (‘‘boolean’’)
       variable.  For example, --follow-ftp tells Wget to follow FTP links
       from HTML files and, on the other hand, --no-glob tells it not to per‐
       form file globbing on FTP URLs.  A boolean option is either affirmative
       or negative (beginning with --no).  All such options share several
       properties.

       Unless stated otherwise, it is assumed that the default behavior is the
       opposite of what the option accomplishes.  For example, the documented
       existence of --follow-ftp assumes that the default is to not follow FTP
       links from HTML pages.

       Affirmative options can be negated by prepending the --no- to the
       option name; negative options can be negated by omitting the --no- pre‐
       fix.  This might seem superfluous---if the default for an affirmative
       option is to not do something, then why provide a way to explicitly
       turn it off?  But the startup file may in fact change the default.  For
       instance, using "follow_ftp = off" in .wgetrc makes Wget not follow FTP
       links by default, and using --no-follow-ftp is the only way to restore
       the factory default from the command line.

       Basic Startup Options


       -V
       --version
           Display the version of Wget.

       -h
       --help
           Print a help message describing all of Wget’s command-line options.

       -b
       --background
           Go to background immediately after startup.  If no output file is
           specified via the -o, output is redirected to wget-log.

       -e command
       --execute command
           Execute command as if it were a part of .wgetrc.  A command thus
           invoked will be executed after the commands in .wgetrc, thus taking
           precedence over them.  If you need to specify more than one wgetrc
           command, use multiple instances of -e.

       Logging and Input File Options


       -o logfile
       --output-file=logfile
           Log all messages to logfile.  The messages are normally reported to
           standard error.

       -a logfile
       --append-output=logfile
           Append to logfile.  This is the same as -o, only it appends to log‐
           file instead of overwriting the old log file.  If logfile does not
           exist, a new file is created.

       -d
       --debug
           Turn on debug output, meaning various information important to the
           developers of Wget if it does not work properly.  Your system
           administrator may have chosen to compile Wget without debug sup‐
           port, in which case -d will not work.  Please note that compiling
           with debug support is always safe---Wget compiled with the debug
           support will not print any debug info unless requested with -d.

       -q
       --quiet
           Turn off Wget’s output.

       -v
       --verbose
           Turn on verbose output, with all the available data.  The default
           output is verbose.

       -nv
       --no-verbose
           Turn off verbose without being completely quiet (use -q for that),
           which means that error messages and basic information still get
           printed.

       -i file
       --input-file=file
           Read URLs from file.  If - is specified as file, URLs are read from
           the standard input.  (Use ./- to read from a file literally named
           -.)

           If this function is used, no URLs need be present on the command
           line.  If there are URLs both on the command line and in an input
           file, those on the command lines will be the first ones to be
           retrieved.  The file need not be an HTML document (but no harm if
           it is)---it is enough if the URLs are just listed sequentially.

           However, if you specify --force-html, the document will be regarded
           as html.  In that case you may have problems with relative links,
           which you can solve either by adding "<base href="url">" to the
           documents or by specifying --base=url on the command line.

       -F
       --force-html
           When input is read from a file, force it to be treated as an HTML
           file.  This enables you to retrieve relative links from existing
           HTML files on your local disk, by adding "<base href="url">" to
           HTML, or using the --base command-line option.

       -B URL
       --base=URL
           Prepends URL to relative links read from the file specified with
           the -i option.

       Download Options


       --bind-address=ADDRESS
           When making client TCP/IP connections, bind to ADDRESS on the local
           machine.  ADDRESS may be specified as a hostname or IP address.
           This option can be useful if your machine is bound to multiple IPs.

       -t number
       --tries=number
           Set number of retries to number.  Specify 0 or inf for infinite
           retrying.  The default is to retry 20 times, with the exception of
           fatal errors like ‘‘connection refused’’ or ‘‘not found’’ (404),
           which are not retried.

       -O file
       --output-document=file
           The documents will not be written to the appropriate files, but all
           will be concatenated together and written to file.  If - is used as
           file, documents will be printed to standard output, disabling link
           conversion.  (Use ./- to print to a file literally named -.)

           Note that a combination with -k is only well-defined for download‐
           ing a single document.

       -nc
       --no-clobber
           If a file is downloaded more than once in the same directory,
           Wget’s behavior depends on a few options, including -nc.  In cer‐
           tain cases, the local file will be clobbered, or overwritten, upon
           repeated download.  In other cases it will be preserved.

           When running Wget without -N, -nc, or -r, downloading the same file
           in the same directory will result in the original copy of file
           being preserved and the second copy being named file.1.  If that
           file is downloaded yet again, the third copy will be named file.2,
           and so on.  When -nc is specified, this behavior is suppressed, and
           Wget will refuse to download newer copies of file.  Therefore,
           ‘‘"no-clobber"’’ is actually a misnomer in this mode---it’s not
           clobbering that’s prevented (as the numeric suffixes were already
           preventing clobbering), but rather the multiple version saving
           that’s prevented.

           When running Wget with -r, but without -N or -nc, re-downloading a
           file will result in the new copy simply overwriting the old.
           Adding -nc will prevent this behavior, instead causing the original
           version to be preserved and any newer copies on the server to be
           ignored.

           When running Wget with -N, with or without -r, the decision as to
           whether or not to download a newer copy of a file depends on the
           local and remote timestamp and size of the file.  -nc may not be
           specified at the same time as -N.

           Note that when -nc is specified, files with the suffixes .html or
           .htm will be loaded from the local disk and parsed as if they had
           been retrieved from the Web.

       -c
       --continue
           Continue getting a partially-downloaded file.  This is useful when
           you want to finish up a download started by a previous instance of
           Wget, or by another program.  For instance:

                   wget -c ftp://sunsite.doc.ic.ac.uk/ls-lR.Z

           If there is a file named ls-lR.Z in the current directory, Wget
           will assume that it is the first portion of the remote file, and
           will ask the server to continue the retrieval from an offset equal
           to the length of the local file.

           Note that you don’t need to specify this option if you just want
           the current invocation of Wget to retry downloading a file should
           the connection be lost midway through.  This is the default behav‐
           ior.  -c only affects resumption of downloads started prior to this
           invocation of Wget, and whose local files are still sitting around.

           Without -c, the previous example would just download the remote
           file to ls-lR.Z.1, leaving the truncated ls-lR.Z file alone.

           Beginning with Wget 1.7, if you use -c on a non-empty file, and it
           turns out that the server does not support continued downloading,
           Wget will refuse to start the download from scratch, which would
           effectively ruin existing contents.  If you really want the down‐
           load to start from scratch, remove the file.

           Also beginning with Wget 1.7, if you use -c on a file which is of
           equal size as the one on the server, Wget will refuse to download
           the file and print an explanatory message.  The same happens when
           the file is smaller on the server than locally (presumably because
           it was changed on the server since your last download
           attempt)---because ‘‘continuing’’ is not meaningful, no download
           occurs.

           On the other side of the coin, while using -c, any file that’s big‐
           ger on the server than locally will be considered an incomplete
           download and only "(length(remote) - length(local))" bytes will be
           downloaded and tacked onto the end of the local file.  This behav‐
           ior can be desirable in certain cases---for instance, you can use
           wget -c to download just the new portion that’s been appended to a
           data collection or log file.

           However, if the file is bigger on the server because it’s been
           changed, as opposed to just appended to, you’ll end up with a gar‐
           bled file.  Wget has no way of verifying that the local file is
           really a valid prefix of the remote file.  You need to be espe‐
           cially careful of this when using -c in conjunction with -r, since
           every file will be considered as an "incomplete download" candi‐
           date.

           Another instance where you’ll get a garbled file if you try to use
           -c is if you have a lame HTTP proxy that inserts a ‘‘transfer
           interrupted’’ string into the local file.  In the future a ‘‘roll‐
           back’’ option may be added to deal with this case.

           Note that -c only works with FTP servers and with HTTP servers that
           support the "Range" header.

       --progress=type
           Select the type of the progress indicator you wish to use.  Legal
           indicators are ‘‘dot’’ and ‘‘bar’’.

           The ‘‘bar’’ indicator is used by default.  It draws an ASCII
           progress bar graphics (a.k.a ‘‘thermometer’’ display) indicating
           the status of retrieval.  If the output is not a TTY, the ‘‘dot’’
           bar will be used by default.

           Use --progress=dot to switch to the ‘‘dot’’ display.  It traces the
           retrieval by printing dots on the screen, each dot representing a
           fixed amount of downloaded data.

           When using the dotted retrieval, you may also set the style by
           specifying the type as dot:style.  Different styles assign differ‐
           ent meaning to one dot.  With the "default" style each dot repre‐
           sents 1K, there are ten dots in a cluster and 50 dots in a line.
           The "binary" style has a more ‘‘computer’’-like orientation---8K
           dots, 16-dots clusters and 48 dots per line (which makes for 384K
           lines).  The "mega" style is suitable for downloading very large
           files---each dot represents 64K retrieved, there are eight dots in
           a cluster, and 48 dots on each line (so each line contains 3M).

           Note that you can set the default style using the "progress" com‐
           mand in .wgetrc.  That setting may be overridden from the command
           line.  The exception is that, when the output is not a TTY, the
           ‘‘dot’’ progress will be favored over ‘‘bar’’.  To force the bar
           output, use --progress=bar:force.

       -N
       --timestamping
           Turn on time-stamping.

       -S
       --server-response
           Print the headers sent by HTTP servers and responses sent by FTP
           servers.

       --spider
           When invoked with this option, Wget will behave as a Web spider,
           which means that it will not download the pages, just check that
           they are there.  For example, you can use Wget to check your book‐
           marks:

                   wget --spider --force-html -i bookmarks.html

           This feature needs much more work for Wget to get close to the
           functionality of real web spiders.

       -T seconds
       --timeout=seconds
           Set the network timeout to seconds seconds.  This is equivalent to
           specifying --dns-timeout, --connect-timeout, and --read-timeout,
           all at the same time.

           When interacting with the network, Wget can check for timeout and
           abort the operation if it takes too long.  This prevents anomalies
           like hanging reads and infinite connects.  The only timeout enabled
           by default is a 900-second read timeout.  Setting a timeout to 0
           disables it altogether.  Unless you know what you are doing, it is
           best not to change the default timeout settings.

           All timeout-related options accept decimal values, as well as sub‐
           second values.  For example, 0.1 seconds is a legal (though unwise)
           choice of timeout.  Subsecond timeouts are useful for checking
           server response times or for testing network latency.

       --dns-timeout=seconds
           Set the DNS lookup timeout to seconds seconds.  DNS lookups that
           don’t complete within the specified time will fail.  By default,
           there is no timeout on DNS lookups, other than that implemented by
           system libraries.

       --connect-timeout=seconds
           Set the connect timeout to seconds seconds.  TCP connections that
           take longer to establish will be aborted.  By default, there is no
           connect timeout, other than that implemented by system libraries.

       --read-timeout=seconds
           Set the read (and write) timeout to seconds seconds.  The ‘‘time’’
           of this timeout refers idle time: if, at any point in the download,
           no data is received for more than the specified number of seconds,
           reading fails and the download is restarted.  This option does not
           directly affect the duration of the entire download.

           Of course, the remote server may choose to terminate the connection
           sooner than this option requires.  The default read timeout is 900
           seconds.

       --limit-rate=amount
           Limit the download speed to amount bytes per second.  Amount may be
           expressed in bytes, kilobytes with the k suffix, or megabytes with
           the m suffix.  For example, --limit-rate=20k will limit the
           retrieval rate to 20KB/s.  This is useful when, for whatever rea‐
           son, you don’t want Wget to consume the entire available bandwidth.

           This option allows the use of decimal numbers, usually in conjunc‐
           tion with power suffixes; for example, --limit-rate=2.5k is a legal
           value.

           Note that Wget implements the limiting by sleeping the appropriate
           amount of time after a network read that took less time than speci‐
           fied by the rate.  Eventually this strategy causes the TCP transfer
           to slow down to approximately the specified rate.  However, it may
           take some time for this balance to be achieved, so don’t be sur‐
           prised if limiting the rate doesn’t work well with very small
           files.

       -w seconds
       --wait=seconds
           Wait the specified number of seconds between the retrievals.  Use
           of this option is recommended, as it lightens the server load by
           making the requests less frequent.  Instead of in seconds, the time
           can be specified in minutes using the "m" suffix, in hours using
           "h" suffix, or in days using "d" suffix.

           Specifying a large value for this option is useful if the network
           or the destination host is down, so that Wget can wait long enough
           to reasonably expect the network error to be fixed before the
           retry.  The waiting interval specified by this function is influ‐
           enced by "--random-wait", which see.

       --waitretry=seconds
           If you don’t want Wget to wait between every retrieval, but only
           between retries of failed downloads, you can use this option.  Wget
           will use linear backoff, waiting 1 second after the first failure
           on a given file, then waiting 2 seconds after the second failure on
           that file, up to the maximum number of seconds you specify.  There‐
           fore, a value of 10 will actually make Wget wait up to (1 + 2 + ...
           + 10) = 55 seconds per file.

           Note that this option is turned on by default in the global wgetrc
           file.

       --random-wait
           Some web sites may perform log analysis to identify retrieval pro‐
           grams such as Wget by looking for statistically significant simi‐
           larities in the time between requests. This option causes the time
           between requests to vary between 0.5 and 1.5 * wait seconds, where
           wait was specified using the --wait option, in order to mask Wget’s
           presence from such analysis.

           A 2001 article in a publication devoted to development on a popular
           consumer platform provided code to perform this analysis on the
           fly.  Its author suggested blocking at the class C address level to
           ensure automated retrieval programs were blocked despite changing
           DHCP-supplied addresses.

           The --random-wait option was inspired by this ill-advised recommen‐
           dation to block many unrelated users from a web site due to the
           actions of one.

       --no-proxy
           Don’t use proxies, even if the appropriate *_proxy environment
           variable is defined.

           For more information about the use of proxies with Wget,

       -Q quota
       --quota=quota
           Specify download quota for automatic retrievals.  The value can be
           specified in bytes (default), kilobytes (with k suffix), or
           megabytes (with m suffix).

           Note that quota will never affect downloading a single file.  So if
           you specify wget -Q10k ftp://wuarchive.wustl.edu/ls-lR.gz, all of
           the ls-lR.gz will be downloaded.  The same goes even when several
           URLs are specified on the command-line.  However, quota is
           respected when retrieving either recursively, or from an input
           file.  Thus you may safely type wget -Q2m -i sites---download will
           be aborted when the quota is exceeded.

           Setting quota to 0 or to inf unlimits the download quota.

       --no-dns-cache
           Turn off caching of DNS lookups.  Normally, Wget remembers the IP
           addresses it looked up from DNS so it doesn’t have to repeatedly
           contact the DNS server for the same (typically small) set of hosts
           it retrieves from.  This cache exists in memory only; a new Wget
           run will contact DNS again.

           However, it has been reported that in some situations it is not
           desirable to cache host names, even for the duration of a short-
           running application like Wget.  With this option Wget issues a new
           DNS lookup (more precisely, a new call to "gethostbyname" or
           "getaddrinfo") each time it makes a new connection.  Please note
           that this option will not affect caching that might be performed by
           the resolving library or by an external caching layer, such as
           NSCD.

           If you don’t understand exactly what this option does, you probably
           won’t need it.

       --restrict-file-names=mode
           Change which characters found in remote URLs may show up in local
           file names generated from those URLs.  Characters that are
           restricted by this option are escaped, i.e. replaced with %HH,
           where HH is the hexadecimal number that corresponds to the
           restricted character.

           By default, Wget escapes the characters that are not valid as part
           of file names on your operating system, as well as control charac‐
           ters that are typically unprintable.  This option is useful for
           changing these defaults, either because you are downloading to a
           non-native partition, or because you want to disable escaping of
           the control characters.

           When mode is set to ‘‘unix’’, Wget escapes the character / and the
           control characters in the ranges 0--31 and 128--159.  This is the
           default on Unix-like OS’es.

           When mode is set to ‘‘windows’’, Wget escapes the characters \, |,
           /, :, ?, ", *, <, >, and the control characters in the ranges 0--31
           and 128--159.  In addition to this, Wget in Windows mode uses +
           instead of : to separate host and port in local file names, and
           uses @ instead of ? to separate the query portion of the file name
           from the rest.  Therefore, a URL that would be saved as
           www.xemacs.org:4300/search.pl?input=blah in Unix mode would be
           saved as www.xemacs.org+4300/search.pl@input=blah in Windows mode.
           This mode is the default on Windows.

           If you append ,nocontrol to the mode, as in unix,nocontrol, escap‐
           ing of the control characters is also switched off.  You can use
           --restrict-file-names=nocontrol to turn off escaping of control
           characters without affecting the choice of the OS to use as file
           name restriction mode.

       -4
       --inet4-only
       -6
       --inet6-only
           Force connecting to IPv4 or IPv6 addresses.  With --inet4-only or
           -4, Wget will only connect to IPv4 hosts, ignoring AAAA records in
           DNS, and refusing to connect to IPv6 addresses specified in URLs.
           Conversely, with --inet6-only or -6, Wget will only connect to IPv6
           hosts and ignore A records and IPv4 addresses.

           Neither options should be needed normally.  By default, an
           IPv6-aware Wget will use the address family specified by the host’s
           DNS record.  If the DNS responds with both IPv4 and IPv6 addresses,
           Wget will try them in sequence until it finds one it can connect
           to.  (Also see "--prefer-family" option described below.)

           These options can be used to deliberately force the use of IPv4 or
           IPv6 address families on dual family systems, usually to aid debug‐
           ging or to deal with broken network configuration.  Only one of
           --inet6-only and --inet4-only may be specified at the same time.
           Neither option is available in Wget compiled without IPv6 support.

       --prefer-family=IPv4/IPv6/none
           When given a choice of several addresses, connect to the addresses
           with specified address family first.  IPv4 addresses are preferred
           by default.

           This avoids spurious errors and connect attempts when accessing
           hosts that resolve to both IPv6 and IPv4 addresses from IPv4 net‐
           works.  For example, www.kame.net resolves to
           2001:200:0:8002:203:47ff:fea5:3085 and to 203.178.141.194.  When
           the preferred family is "IPv4", the IPv4 address is used first;
           when the preferred family is "IPv6", the IPv6 address is used
           first; if the specified value is "none", the address order returned
           by DNS is used without change.

           Unlike -4 and -6, this option doesn’t inhibit access to any address
           family, it only changes the order in which the addresses are
           accessed.  Also note that the reordering performed by this option
           is stable---it doesn’t affect order of addresses of the same fam‐
           ily.  That is, the relative order of all IPv4 addresses and of all
           IPv6 addresses remains intact in all cases.

       --retry-connrefused
           Consider ‘‘connection refused’’ a transient error and try again.
           Normally Wget gives up on a URL when it is unable to connect to the

lib/imw/rip/wget.rb
#
# h2. imw/rip/wget.rb -- Interface to 'wget' utility
#
# == Downloading Data
#
# These functions act as an interface to the popular 'wget' utility
# for downloading data from Web resources.
#
# Other options to try:
#
# * To --accept only files with these extensions (-R to reject):
#     -A.jpg,.jpeg,.png,.gif,.htm,.html,.[xs]html,.x[sm][dl],.php,.as[px],.cgi,.pl
#   (can also force an --html-extension -k -K and you can
#   and --span-hosts --domains=... OR --exclude-domains=...
#   and --include-directories=...  OR --exclude-directories=...)
# * To --append-output log; first is brief, second gives 1M per line.
#     -nv -a wget-`date +%Y%m%d`.log     
#     -S  -a wget-`date +%Y%m%d`.log --debug --verbose --progress=dot -edot_bytes=32k -edots_in_line=32
# * To be polite:
#     --limit-rate=25k -w0.5 --tries=3 --header='From: Your Name <username@site.domain>' 
# * To be impolite:
#     --no-cache -erobots=off
# * To --convert-links to be local, getting all --page_requisites; (you 
#   can add -k -K to save the .orig inal)
#     -k -p		
# * To use --timestamping instead of just -nc for --no-clobber ; if you
#   do this *and* --convert-links you should also --backup-converted
#     -N --no-remove-listing 			
# * List of files to use:
#     --input-file=file --base=URL
#   (instead of text you can tell it it's a --force_html file)
# * Filename munging: translate a lot OR translate minimally
#   --restrict-file-names=windows     OR --restrict-file-names=unix,nocontrol
# * Chop hostname and n directory levels OR make --no-directories at all
#   -nH --cut-dirs=n  	     
#   -nd -P DIR-I-WANT-INSTEAD
#
# These can solve some problems:
#   --debug --verbose -S
#   --server-response shows headers in log, --save-headers keeps with file   
#   --post_file=file    --post-data=string	
#   --no-cache 		--no-dns-cache		--ignore_length 
#   --referer=string 	-S --password --user	--header "Cookie: name=value"
#   --user-agent="Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.8.1.4) Gecko/20061201 Firefox/2.0.0.4 (Ubuntu-feisty)"
#     or --user-agent="" for no user agent sent
#   
# Log in then fetch:
#   wget --save-cookies cookies.txt --keep-session-cookies --post-data 'user=foo&password=bar' <loginurl>
#   wget --load-cookies cookies.txt --keep-session-cookies <fileurl>
#
# SSL:
#   --secure-protocol=auto --no-check-certificate
#
# Also, add these to your ~/.wgetrc (command line always overrides):
#   follow_ftp 	 	 = on
#   check-certificate    = off
#   # if your conscience so allows
#   robots		 = off	
#   # cache		 = off
#   # timestamping 	 = on
#   # no-clobber 	 = on	 
#   # do ln -s or cp -p ~/.firefox/asdf/cookies.txt ~/.wget/cookies.txt
#   cookies		 = on    
#   keep-session-cookies = on
#   load_cookies	 = /home/flip/.wget/cookies.txt
#   save_cookies	 = /home/flip/.wget/cookies.txt
#   user-agent="Mozilla/5.0 (X11; U; Linux i686; en-US; rv:1.8.1.4) Gecko/20061201 Firefox/2.0.0.4 (Ubuntu-feisty)" 
#
# Simpler form:
#   wget -r -l5 -nc -np -nv -b -a wget.log -erobots=off -w1 $@
# 
# turn off 'background'
# mandatory: --no-clobber --no-parent --no-verbose
# options (upt o user): -r, -l, url, -limit-rate, -wait
#     * if not recursive then use -x (full path)
#     * also look at -P and -nH (for control over directories)
# make sure to (after the -a flag) specify a log in the 'dump
# section of each pool (by using the $imw.path_to(:dump,'url.log')
# think about what the end user will want to input:
# -url
# -fields
# -update frequency
# and feed.rb should do the rest
# wget -r -l5  --no-clobber --no-parent                     	\
#     --no-verbose --background \
#     -a ~/slice/var/log/wget-sitescrape-`date +%Y%m%d`.log 	\
#     --wait=0.5 --random-wait --limit-rate=100k	\
#     $@
#
# Author::    (Philip flip Kromer, Dhruv Bansal) for Infinite Monkeywrench Project (mailto:coders@infochimps.org)
# Copyright:: Copyright (c) 2008 infochimps.org
# License::   GPL 3.0
# Website::   http://infinitemonkeywrench.org/
# 

require 'imw'
require 'uri'

$imw = IMW.new_from_env() # should we be doing things this way?

# Download pages using the 'wget' utility.  The optional argument
# 'wget_path' specifies the path to the wget binary.
#
# Optional arguments to 'wget' can be provided through the 'options'
# hash which should consist of values keyed to the strings below along
# with the 'wget' flags they correspond to (parenthesis) and the
# default values (brackets).  See the man pages for 'wget' for more
# information.
# 
#  * :directory (-P) [ripd directory]
#  * :max_tries (-t) [20]
#  * :background (-b) [false]
#  * :recursive (-r) [true]
#  * :span_hosts (-H) [false]
#  * :no_directories (-nd) [false]
#  * :max_depth (-l) [5]
#  * :no_parent (-np) [true]
#  * :accept_extensions (-A) [nil]
#  * :reject_extensions (-R) [nil]
#  * :no_clobber (-nc) [true]
#  * :verbosity (-v,-nv,-q) [q]
#  * :output_file (-O) [nil]
#  * :log_file (-a) ['url.log' in dump directory]
#  * :continue (-c) [false]
#  * :wait (-w) [0.5]
#  * :random_wait (--random-wait) [true]
#  * :limit_rate (--limit-rate) [nil]
#
# Alternatively, 'options' can just be a string consisting of options
# to feed to wget directly.

Default_options = {:max_tries=>20,  :recursive=>true,:no_directories=>false,:max_depth=>5,:no_parent=>true,:no_clobber=>false,:verbosity=>'q',:continue=>false,:wait=>0.5,:random_wait=>true,:span_hosts=>false}

def wget(uri,wget_path='wget',options=nil)
  # get path to name logfile with
  uri_object = URI.parse(uri)
  path = (uri_object.host + uri_object.path).gsub('/','_') + '.' + Time.now.strftime("%Y-%M-%d_%H:%M:%S")

  if not options then options = Default_options end

  # make sure to check that no control characters are passed to wget
  cc_regex = /(#|;|\$)/
  mesg = "Control character (`#', `;', `$') found in options."
  if options.class == String then raise ArgumentError, mesg if options =~ cc_regex
  elsif options.class == Hash then
    options.each_value { |option| raise ArgumentError, mesg if option =~ cc_regex }
  end

  # clean some options
  options[:verbosity] = options[:verbosity].gsub('-','')

  # assemble command
  command = "#{wget_path} "
  if options.class == String then command += options
  elsif options.class == Hash then
    command += "-t#{options[:max_tries]} -l#{options[:max_depth]} -#{options[:verbosity]} -w#{options[:wait]} "
    flags = []
    if options[:span_hosts] then flags << "-H" end
    if options[:background] then flags << "-b" end
    if options[:recursive] then flags << "-r" end
    if options[:no_directories] then flags << "-nw" end
    if options[:no_parent] then flags << "-np" end
    if options[:no_clobber] then flags << "-nc" end
    if options[:output_file] then flags << "-O#{options[:output_file]}" end
    if options[:log_file] then flags << "-a#{options[:log_file]}" else flags << "-a#{$imw.path_to(:dump,path + ".log")}" end
    if options[:continue] then flags << "-c" end
    if options[:random_wait] then flags << "--random-wait" end
    if options[:limit_rate] then flags << "--limit-rate=#{options[:limit_rate]}" end
    if options[:accept_extensions] then flags << "-A#{options[:accept_extensions]}" end
    if options[:reject_extensions] then flags << "-R#{options[:reject_extensions]}" end
    if options[:directory] then flags << "-P#{options[:directory]}" else flags << "-P#{$imw.path_to(:ripd)}" end
    command += flags.join(' ')
  end
  puts command
end



puts "#{File.basename(__FILE__)}: You walk through the trees jauntily clubbing datasets with your Monkeywrench." # at bottom
